### PR TITLE
refactor(clickhouse): apply repetitive transformations as pattern replacements

### PR DESCRIPTION
**Pull request by cpcloud**
Use a pattern for repetitive transformations like nullifying strings and subtracting one from more one-based indexed operations

### CODE REVIEW

File: ibis/backends/clickhouse/compiler/core.py [hunk 1]
```diff
@@ -113,22 +113,30 @@ def fn(node, _, **kwargs):
...
+
     op = op.replace(
         replace_literals
         | replace_in_column_with_table_array_view
         | replace_empty_in_values_with_false
         | replace_notexists_subquery_with_not_exists
         | replace_notany_with_min_not
         | replace_notall_with_max_not
-        | subtract_one_from_ranking_functions
+        | subtract_one_from_one_based_functions
```
Comment (cpcloud) [inline ibis/backends/clickhouse/compiler/core.py:137]: Rename to `subtract_one_from_one_indexed_functions`

**Comment by kszucs** [comment]
What do you think, are these going to be reusable? Soon enough we should have simple unit tests for the "named" rewrite rules to spare some debugging time in the future.

**Comment by cpcloud** [comment]
Some of these rules are reusable and some are specific to clickhouse:

### Reusable

- `replace_literals`
- `replace_in_column_with_table_array_view`
- `replace_empty_in_values_with_false`
- `subtract_one_from_one_indexed_functions`
- `add_one_to_nth_value_input`

### Specific to ClickHouse

- `nullify_empty_string_results`

File: ibis/backends/clickhouse/compiler/values.py [hunk 2]
```diff
@@ -597,7 +592,8 @@ def _in_column(op, *, value, options, **_):
     return value.isin(options.this if isinstance(options, sg.exp.Subquery) else options)
 
 
-_NUM_WEEKDAYS = 7
+_DAYS = calendar.day_name
+_NUM_WEEKDAYS = len(_DAYS)
```
Comment (kszucs) [inline ibis/backends/clickhouse/compiler/values.py:596]: Could we use just `7` everywhere? Currently I have to track 3 variables whereas it is always just `7`.

```py
weekdays = _NUM_WEEKDAYS
_NUM_WEEKDAYS = len(_DAYS)
_DAYS = calendar.day_name
```

Comment (kszucs) [reply ibis/backends/clickhouse/compiler/values.py:596]: I'd prefer to have `_NUM_WEEKDAYS` removed entirely, can be done in a follow-up.

**Review by kszucs — APPROVED** [review]
Nice refactor!

File: ibis/backends/clickhouse/compiler/values.py [hunk 3]
```diff
@@ -804,6 +796,16 @@ def formatter(op, *, left, right, **_):
...
+    ops.NthValue: "nth_value",
+    ops.MinRank: "rank",
+    ops.DenseRank: "dense_rank",
+    ops.RowNumber: "row_number",
+    ops.ExtractProtocol: "protocol",
+    ops.ExtractAuthority: "netloc",
+    ops.ExtractHost: "domain",
+    ops.ExtractPath: "path",
+    ops.ExtractFragment: "fragment",
+    ops.ArrayPosition: "indexOf",
```
Comment (kszucs) [inline ibis/backends/clickhouse/compiler/values.py:808]: Pretty nice to see this simple mapping!

**Pull request merged**